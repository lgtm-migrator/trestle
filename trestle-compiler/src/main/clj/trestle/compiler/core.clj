(ns trestle.compiler.core
  (:require [clojure.spec.alpha :as s]
            [cheshire.core :as c]))

(s/def
  ::version
  (s/and pos-int? #(>= 1 %)))

(s/def ::dataset string?)
(s/def ::name string?)
(s/def ::type string?)
(s/def ::spatial boolean?)
(s/def ::identifier boolean?)
(s/def ::temporal string?)

; This isn't working, I want to enforce that we can only have a single qualifier on each fact
; Not sure how to do that
; I'd also like to ensure we have a maximum of 1 ID or spatial value
(s/def ::fact-qualifiers (s/or :spatial ::spatial :identifier ::identifier :temporal ::temporal))

;(s/def ::fact-qualifiers (s/or
;                           (s/keys :req-un [:spatial ::spatial])
;                           (s/keys :req-un [:identifier ::identifier])
;                           (s/keys :req-un [:temporal ::temporal])))

(s/def ::fact-params (s/keys
                       :req-un [::name ::type]
                       :opt-un [::fact-qualifiers]))

(s/def ::fact (s/merge ::fact-qualifiers ::fact-params))

(s/def ::facts (s/coll-of ::fact))

(s/def :dataset/spec
  (s/keys :req-un [::version ::dataset ::facts]))

(defn parse-json [file]
  (c/parse-stream (clojure.java.io/reader file) true))

(s/fdef parse-json
        :args (s/cat :file string?)
        :ret ::dataset)