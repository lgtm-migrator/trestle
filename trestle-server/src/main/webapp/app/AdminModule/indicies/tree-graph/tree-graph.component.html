<h4>I'm a graph!</h4>
<div #graph class="container" style="width: 100%"></div>