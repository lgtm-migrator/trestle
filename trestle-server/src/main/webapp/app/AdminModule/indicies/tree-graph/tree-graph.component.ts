import { AfterViewInit, Component, ElementRef, Input, OnChanges, SimpleChanges, ViewChild } from "@angular/core";
import { ScaleLinear, scaleLinear } from "d3-scale";
import { BaseType, select, Selection } from "d3-selection";
import * as moment from "moment";
import { Moment } from "moment";
import { axisBottom, axisLeft } from "d3-axis";
import { IIndexLeafStatistics } from "../index.service";

@Component({
    selector: "tree-graph",
    templateUrl: "./tree-graph.component.html",
    styleUrls: ["./tree-graph.component.css"]
})
export class TreeGraphComponent implements AfterViewInit, OnChanges {

    @ViewChild("graph")
    public element: ElementRef;
    @Input()
    public data: IIndexLeafStatistics[];
    private htmlElement: HTMLElement;
    private host: Selection<HTMLElement, IIndexLeafStatistics, null, undefined>;
    private svg: Selection<BaseType, IIndexLeafStatistics, null, undefined>;
    private width: number;
    private height: number;
    private margin: ID3Margin;
    private maxTime: Moment;
    private x: ScaleLinear<number, number>;
    private y: ScaleLinear<number, number>;

    public constructor() {
        this.maxTime = moment("3001-01-01").startOf("year");
    }

    public ngAfterViewInit(): void {
        this.htmlElement = this.element.nativeElement;
        this.setupD3();
    }

    public ngOnChanges(changes: SimpleChanges): void {
        const dataChanges = changes["data"];
        if (dataChanges.currentValue !== dataChanges.previousValue) {
            console.debug("Plotting new changes");
            this.plotData(dataChanges.currentValue);
        }
    }

    private plotData(data: IIndexLeafStatistics[]): void {
    //    For each, leaf, draw the triangle
        const leafData = this.svg
            .selectAll(".leaf")
            .data(data, (d: any) => d.leafID);

        leafData
            .enter()
            .append("polygon")
            .attr("class", "leaf")
            .attr("points", (d: any) => this.normalizeTriangle(d.coordinates))
            .merge(leafData);
    }

    /**
     * Normalize coordinates (represented as millis from UTC epoch) into D3 coordinates
     * @param {number[]} coordinates
     * @returns {string}
     */
    private normalizeTriangle(coordinates: number[]): string {
        const size = coordinates.length;
        const normalized = new Array(size);
        let isX = true;
        for (let idx = 0; idx < size; idx++) {
            normalized[idx] = isX ? this.x(coordinates[idx]) : this.y(coordinates[idx]);
            isX = !isX;
        }
        return normalized.join(",");
    }

    private setupD3(): void {
        this.host = select<HTMLElement, IIndexLeafStatistics>(this.htmlElement);
        this.margin = this.margin = {top: 20, right: 30, bottom: 20, left: 30};
        this.width = this.htmlElement.offsetWidth - this.margin.left - this.margin.right;
        this.height = 500 - this.margin.top - this.margin.bottom;

        // Set the X/Y axis
        this.x = scaleLinear().range([0, this.width]);
        this.x.domain([0, this.maxTime.valueOf()]);
        this.y = scaleLinear().range([0, this.height]);
        this.y.domain([0, this.maxTime.valueOf()]);

        console.debug("Y:", this.y);
        console.debug("Yd:", this.y.domain());
        console.debug("Yr:", this.y.range());

        console.debug("X:", this.x);
        console.debug("Xd:", this.x.domain());
        console.debug("Xr:", this.x.range());

        this.svg = this.host.html("")
            .append("svg")
            .attr("width", this.width + this.margin.left + this.margin.right)
            .attr("height", this.height + this.margin.top + this.margin.bottom)
            .append("g")
            .attr("transform", "translate(" + this.margin.left + "," + this.margin.top + ")");

        // Add the axises
        this.svg
            .append("g")
            .attr("class", "axis x-axis")
            .attr("transform", "translate(0," + this.height + ")")
            .call(axisBottom(this.x));

        this.svg
            .append("g")
            .attr("class", "axis axis-y")
            .append("text")
            .attr("transform", "rotate(-90)")
            .attr("y", 6)
            .attr("dy", "0.71em")
            .attr("fill", "#000")
            .text("Value")
            .call(axisLeft(this.y));

        //    Draw a line
        this.svg
            .append("g")
            .attr("class", "dividing-line")
            .append("line")
            .attr("x1", 0)
            .attr("x2", this.x(this.maxTime.valueOf()))
            // .attr("y1", this.height)
            .attr("y1", this.y(this.maxTime.valueOf()))
            .attr("y2", 0)
            .attr("stroke-width", 2)
            .attr("stroke", "black");
    }
}
