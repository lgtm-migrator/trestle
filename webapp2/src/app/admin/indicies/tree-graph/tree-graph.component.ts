import { AfterViewInit, Component, ElementRef, EventEmitter, Input, OnChanges, Output, SimpleChanges, ViewChild } from "@angular/core";
import { ScaleLinear, scaleLinear } from "d3-scale";
import { BaseType, select, Selection } from "d3-selection";
import moment, { Moment } from "moment";
import { axisBottom, axisLeft } from "d3-axis";
import { IIndexLeafStatistics } from "../index.service";
import { interpolateHsl } from "d3-interpolate";
import {ID3Margin} from '../../../ui/common';
import {BehaviorSubject, Subject} from 'rxjs';
import {RoundingPipe} from '../../../shared/pipes/rounding-pipe.pipe';

export interface IGraphHeader {
    maxValue: number;
    offsetValue: number;
    leafs: IIndexLeafStatistics[];
}

@Component({
    selector: "tree-graph",
    templateUrl: "./tree-graph.component.html",
    styleUrls: ["./tree-graph.component.scss"]
})
export class TreeGraphComponent implements AfterViewInit, OnChanges {

    @ViewChild("graph")
    public element: ElementRef;
    @Input()
    public data: IGraphHeader;
    @Output()
    public hovered: EventEmitter<string>;
    private htmlElement: HTMLElement;
    private host: Selection<HTMLElement, IIndexLeafStatistics, null, undefined>;
    private svg: Selection<BaseType, IIndexLeafStatistics, null, undefined>;
    private width: number;
    private height: number;
    private margin: ID3Margin;
    private maxTime: Moment;
    private x: ScaleLinear<number, number>;
    private y: ScaleLinear<number, number>;
    private colorScale: (value: number) => string;
    private dataSubject: Subject<IGraphHeader | undefined>;
    private rounder: RoundingPipe;

    public constructor() {
        this.maxTime = moment("5000-01-01").startOf("year");
        this.colorScale = interpolateHsl("steelblue", "brown");
        this.rounder = new RoundingPipe();
        this.hovered = new EventEmitter<string>();
        this.dataSubject = new BehaviorSubject(undefined);
    }

    public ngAfterViewInit(): void {
        this.htmlElement = this.element.nativeElement;
        this.setupD3();

        //    Subscribe
        this.dataSubject
            .subscribe((value) => {
                if (value !== undefined) {
                    this.plotData(value);
                    this.data = value;
                }
            });
    }

    public ngOnChanges(changes: SimpleChanges): void {
        const dataChanges = changes["data"];
        if (dataChanges.currentValue !== dataChanges.previousValue) {
            console.debug("Plotting new changes");
            this.dataSubject.next(dataChanges.currentValue);
        }
    }

    private plotData(data: IGraphHeader): void {
        // Nuke everything, because I can't figure out the update pattern
        this.svg.selectAll("*").remove();
        // Calculate the max time, which need to adjust from the cache values
        this.maxTime = moment(data.maxValue - data.offsetValue);

        console.debug("Plotting with max: %s and offset: %s", data.maxValue, data.offsetValue);

        // Set the X/Y scales
        this.x = scaleLinear().range([0, this.width]);
        this.x.domain([0, data.maxValue]);
        // We need to invert this, in order to get the triangles to draw in the correct direction
        this.y = scaleLinear().range([this.height, 0]);
        this.y.domain([0, data.maxValue]);

        // Add the axises
        this.svg
            .append("g")
            .attr("class", "axis x-axis")
            .attr("transform", "translate(0," + this.height + ")")
            .call(axisBottom(this.x)
                .tickFormat((d) => TreeGraphComponent.adjustTemporals(d, data.offsetValue)))
            .selectAll("text")
            .attr("y", 0)
            .attr("x", 9)
            .attr("dy", ".35em")
            .attr("transform", "rotate(45)")
            .style("text-anchor", "start");

        this.svg
            .append("g")
            .attr("class", "axis axis-y")
            .call(axisLeft(this.y)
                .tickFormat((d) => TreeGraphComponent.adjustTemporals(d, data.offsetValue)))
            .selectAll("text")
            .attr("y", 0)
            .attr("x", -10)
            .attr("dy", ".35em")
            .style("text-anchor", "end");

        //    Draw a line
        // this.svg
        //     .append("g")
        //     .attr("class", "dividing-line")
        //     .append("line")
        //     .attr("x1", 0)
        //     .attr("x2", this.x(data.maxValue))
        //     // .attr("y1", this.height)
        //     .attr("y1", this.height)
        //     .attr("y2", 0)
        //     .attr("stroke-width", 3)
        //     .attr("stroke", "black");


        //    For each, leaf, draw the triangle
        const leafData = this.svg
            .selectAll<SVGPolygonElement, BaseType>(".leaf")
            // This index function needs to be a string, for some reason
            .data(data.leafs, (d: IIndexLeafStatistics) => d.binaryID);

        leafData
            .enter()
            .append("polygon")
            .attr("id", (d) => d.binaryID)
            .attr("direction", (d) => d.direction.toString())
            .attr("coords", (d) => d.coordinates.join(","))
            .attr("class", "leaf")
            .attr("points", (d) => this.normalizeTriangle(d.coordinates))
            .attr("stroke-width", 1)
            .attr("stroke", "black")
            // .attr("fill", (d) => this.colorScale((d.records / 20) + 0.01))
            .attr("fill", "blue")
            .style("fill-opacity", 0.7)
            // Hover handlers
            .on("mouseover", this.hoverHandler)
            .merge(leafData);

    }

    private hoverHandler = (d: IIndexLeafStatistics): void => {
        console.debug("Hovered", d.leafID.toString());
        this.hovered.next(d.leafID.toString());
    };

    /**
     * Normalize coordinates (represented as millis from UTC epoch) into D3 coordinates
     * @param {number[]} coordinates
     * @returns {string}
     */
    private normalizeTriangle(coordinates: number[]): string {
        const size = coordinates.length;
        const normalized = new Array(size);
        let isX = true;
        for (let idx = 0; idx < size; idx++) {
            // console.debug("Rounding %s to %s", coordinates[idx], this.rounder.transform(coordinates[idx], 0));
            const rounded = this.rounder.transform(coordinates[idx], 0);
            normalized[idx] = isX ? this.x(rounded) : this.y(rounded);
            isX = !isX;
        }
        return normalized.join(",");
    }

    private static adjustTemporals(domain: any, offsetValue: number): string {
        return new Date(domain - offsetValue).toISOString();
    }

    private setupD3(): void {
        this.host = select<HTMLElement, IIndexLeafStatistics>(this.htmlElement);
        this.margin = this.margin = {top: 20, right: 30, bottom: 100, left: 150};
        this.width = this.htmlElement.offsetWidth - this.margin.left - this.margin.right;
        this.height = 500 - this.margin.top - this.margin.bottom;

        this.svg = this.host.html("")
            .append("svg")
            .attr("width", this.width + this.margin.left + this.margin.right)
            .attr("height", this.height + this.margin.top + this.margin.bottom)
            .append("g")
            .attr("transform", "translate(" + this.margin.left + "," + this.margin.top + ")");
    }
}
