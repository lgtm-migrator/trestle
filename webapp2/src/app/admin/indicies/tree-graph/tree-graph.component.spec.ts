import {async, ComponentFixture, TestBed} from '@angular/core/testing';
import {NO_ERRORS_SCHEMA} from '@angular/core';
import {TreeGraphComponent} from './tree-graph.component';

describe('TreeGraphComponent', () => {
  let component: TreeGraphComponent;
  let fixture: ComponentFixture<TreeGraphComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [TreeGraphComponent],
      schemas: [NO_ERRORS_SCHEMA]
    })
      .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(TreeGraphComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toMatchSnapshot();
  });
});
