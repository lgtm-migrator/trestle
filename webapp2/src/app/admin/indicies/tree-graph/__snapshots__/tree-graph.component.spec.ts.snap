// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeGraphComponent should create 1`] = `
TreeGraphComponent {
  "colorScale": [Function],
  "dataSubject": BehaviorSubject {
    "_isScalar": false,
    "_value": undefined,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [
      Subscriber {
        "_parentOrParents": null,
        "_subscriptions": Array [
          SubjectSubscription {
            "_parentOrParents": [Circular],
            "_subscriptions": null,
            "closed": false,
            "subject": [Circular],
            "subscriber": [Circular],
          },
        ],
        "closed": false,
        "destination": SafeSubscriber {
          "_complete": undefined,
          "_context": [Circular],
          "_error": undefined,
          "_next": [Function],
          "_parentOrParents": null,
          "_parentSubscriber": [Circular],
          "_subscriptions": null,
          "closed": false,
          "destination": Object {
            "closed": true,
            "complete": [Function],
            "error": [Function],
            "next": [Function],
          },
          "isStopped": false,
          "syncErrorThrowable": false,
          "syncErrorThrown": false,
          "syncErrorValue": null,
        },
        "isStopped": false,
        "syncErrorThrowable": true,
        "syncErrorThrown": false,
        "syncErrorValue": null,
      },
    ],
    "thrownError": null,
  },
  "element": ElementRef {
    "nativeElement": <div>
      <svg
        height="500"
        width="0"
      >
        <g
          transform="translate(150,20)"
        />
      </svg>
    </div>,
  },
  "height": 380,
  "host": Selection {
    "_groups": Array [
      Array [
        <div>
          <svg
            height="500"
            width="0"
          >
            <g
              transform="translate(150,20)"
            />
          </svg>
        </div>,
      ],
    ],
    "_parents": Array [
      null,
    ],
  },
  "hoverHandler": [Function],
  "hovered": EventEmitter {
    "__isAsync": false,
    "_isScalar": false,
    "closed": false,
    "hasError": false,
    "isStopped": false,
    "observers": Array [],
    "thrownError": null,
  },
  "htmlElement": <div>
    <svg
      height="500"
      width="0"
    >
      <g
        transform="translate(150,20)"
      />
    </svg>
  </div>,
  "margin": Object {
    "bottom": 100,
    "left": 150,
    "right": 30,
    "top": 20,
  },
  "maxTime": "5000-01-01T05:00:00.000Z",
  "rounder": RoundingPipe {},
  "svg": Selection {
    "_groups": Array [
      Array [
        <g
          transform="translate(150,20)"
        />,
      ],
    ],
    "_parents": Array [
      null,
    ],
  },
  "width": -180,
}
`;
